### PR TITLE
Sockets - Cyndi

### PR DESCRIPTION
**Pull request by cyndilopez**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Which classes to implement and how classes communicate with the module.
What was a design decision you made that changed over time over the project? | The behavior of room changed; originally it had no behavior and then was changed to be able to communicate whether it was available.
What was a concept you gained clarity on, or a learning that you'd like to share? | I gained clarity on what guard and coverage do.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | A nominal test was checking whether the code could calculate the total cost of a reservation correctly for a span of a few days. It's nominal because it's a standard calculation with simple math.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | Edge case date ranges were used to test whether a room was available for a given date range. It was an edge case because it was outside the norm.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I should've written more pseudocode; however I did well on first writing tests and then the code.

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      @id = id
+      @rooms = rooms
+      @reservations ||= reservations
+      @hotel_blocks = []
+    end
+
+    def reserve_room(start_date, end_date, room_id, block_id = nil, discount = nil)
+      #check if room available during those days
+      rooms_available = available_rooms(start_date, end_date)
+      raise ArgumentError if !(rooms_available.include?(room_id)) && block_id == nil
```
Comment (droberts-sea) [inline lib/hotel.rb:15]: What if instead of calling `available_rooms` and then checking for inclusion, you called `find_room` and then `is_available?` directly? That would save you calling `is_available?` on all the other rooms.

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    end
+
+    def find_reservation(room_id, block_id)
+      return reservations.detect { |reservation| reservation.room.id == room_id && reservation.block_id == block_id }
+    end
+
+    def available_rooms(start_date, end_date)
+      avail_rooms = []
+      rooms.each do |room|
+        avail_rooms << room.id if room.isAvailable?(start_date, end_date)
```
Comment (droberts-sea) [inline lib/hotel.rb:56]: The `.select` enumerable might be helpful to clean up this code a little.

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    end
+
+    def add_reservation(reservation)
+      reservations << reservation
+    end
+
+    def access_reservations(date)
+      list_reservations = @reservations.select { |reservation| (Range.new(reservation.start_date, reservation.end_date)).include?(date) }
+      return list_reservations.map { |reservation| reservation }
+    end
```
Comment (droberts-sea) [inline lib/hotel.rb:47]: On line 45 you do a lot of work with the internals of the reservation. I know we hadn't read POODR ch 4 yet when you wrote this code, but this is exactly what Metz is talking about when she says you should "ask for _what_ instead of telling _how_". A cleaner approach might be to add a `.includes_date?` method to reservation, then use it here:

```ruby
return @reservations.select { |res| res.include_date?(date) }
```

Also the `map` on line 46 is not needed.

File: lib/reservation.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    def initialize(id:, start_date:, end_date:, room:, total_cost: nil, block_id: nil, discount: nil)
+      @id = id
+      @start_date = start_date
+      @end_date = end_date
+      raise ArgumentError if end_date <= start_date
+      @room = room
+      @block_id = block_id
+      @discount = discount
+      @total_cost = calc_total_cost
+    end
```
Comment (droberts-sea) [inline lib/reservation.rb:14]: In stead of storing `@total_cost` in an instance variable, it might be wise to rename the `calc_total_cost` method `total_cost`. That way if the price, discount, or number of nights were ever to change, the total price would "update" automatically (update is in quotes because it was never stored in the first place).

File: lib/room.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    def isAvailable?(start_date, end_date)
+      return check_date_range_conflict(start_date, end_date, reservations) &&
+               check_date_range_conflict(start_date, end_date, block)
+    end
+
+    def check_date_range_conflict(start_date, end_date, array)
+      if array != []
+        array.each do |item|
+          return true if (!(item.start_date...item.end_date).include?(start_date) &&
```
Comment (droberts-sea) [inline lib/room.rb:29]: I would ask for more descriptive variable names here than `array` and `item`. What do you expect these variables to contain?

File: spec/block_spec.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+describe "Block class" do
+  before do
+    @rooms = []
+    20.times do |k|
+      @rooms << Hotel::Room.new(id: k)
+    end
+    @hotel = Hotel::Hotel.new(
+      id: 1,
+      rooms: @rooms,
```
Comment (droberts-sea) [inline spec/block_spec.rb:12]: Rather than messing around with an entire `Hotel` in the `Block` tests, you should instantiate a `Block` directly and test on that.

File: spec/hotel_spec.rb [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+  describe "#reserve_room" do
+    it "can create a reservation" do
+      new_reservation = @hotel.reserve_room(Date.new(2001, 3, 5), Date.new(2001, 3, 7), 0)
+      expect(new_reservation).must_be_kind_of Hotel::Reservation
+    end
+
+    it "adds the reservation to the Hotel's list of reservations" do
+      new_reservation = @hotel.reserve_room(Date.new(2001, 3, 5), Date.new(2001, 3, 7), 0)
+      expect(@hotel.reservations.length).must_equal 1
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:43]: What happens if you try to reserve a room that's already taken?

File: spec/hotel_spec.rb [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    end
+  end
+
+  describe "#find_room" do
+    it "returns object associated with room id" do
+      expect(@hotel.find_room(0)).must_be_kind_of Hotel::Room
+    end
+
+    it "returns the correct room id" do
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:52]: I would probably combine these two tests into one. However there is a case you're missing: what if the room to be found doesn't exist?

File: spec/hotel_spec.rb [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    it "existing reservation begins and ends reservation" do
+      start_date = Date.new(2001, 3, 7)
+      end_date = Date.new(2001, 3, 10)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "raises an Argument Error if tries to reserve room that is not available for a given day" do
+      expect { @hotel.reserve_room(Date.new(2001, 3, 5), Date.new(2001, 3, 17), 12) }.must_raise ArgumentError
+    end
+  end
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:143]: Aha, here is the test I asked for above. Since this exercises the `reserve_room` method, it should be grouped with those tests.

File: spec/room_spec.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
+require_relative "spec_helper"
+require "Date"
+describe "Room class" do
+  describe "Room instantiation" do
+    before do
```
Comment (droberts-sea) [inline spec/room_spec.rb:5]: You should have some tests for the `is_available?` method.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes - the test case you've labeled as an edge case (checking whether a room is available) is really more of a nominal case. There's not an "edge" in the behavior of the program that the test is exploring. An edge case would be checking if a room is available the day after the previous reservation ends.
**Design** |
Each class is responsible for a single piece of the program |
Classes are loosely coupled | mostly
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | mostly
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | yes
**Fundamentals** |
Names variables, classes and modules appropriately | mostly - see inline
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | mostly - see inline
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes 
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Good work overall! This code is well-written and does a good job of solving the problem. I've left a fair bit of feedback inline, but most of it is for little things, and I am confident that the learning goals for this assignment were met. Keep up the hard work!
